pkg/bgpv1/gobgp/reconcile.go

```diff
@@ -0,0 +1,424 @@
+func preflightReconciler(ctx context.Context, _ *BGPRouterManager, sc *ServerWithConfig, newc *v2alpha1api.CiliumBGPVirtualRouter, cstate *agent.ControlPlaneState) error {
+	var (
+		l = log.WithFields(
+			logrus.Fields{
+				"component": "gobgp.preflightReconciler",
+			},
+		)
+	)
+
+	// If we have no config attached, we don't need to perform a preflight for
+	// reconciliation.
+	//
+	// This is the first time this server is being registered and BGPRouterManager
+	// set any fields needing reconcilation in this function already.
+	if sc.Config == nil {
+		l.Debugf("Preflight for virtual router with ASN %v not necessary, first instantiation of this BgpServer.", newc.LocalASN)
+		return nil
+	}
+
+	l.Debugf("Begin preflight reoncilation for virtual router with ASN %v", newc.LocalASN)
+	bgpInfo, err := sc.Server.GetBgp(ctx, &gobgp.GetBgpRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve BgpServer info for virtual router with ASN %v: %w", newc.LocalASN, err)
+	}
+
+	// resolve local port from kubernetes annotations
+	var localPort int32
+	localPort = -1
+	if attrs, ok := cstate.Annotations[newc.LocalASN]; ok {
+		if attrs.LocalPort != 0 {
+			localPort = int32(attrs.LocalPort)
+		}
+	}
+
+	// resolve router ID, if we have an annotation and it can be parsed into
+	// a valid ipv4 address use this,
+	//
+	// if not determine if Cilium is configured with an IPv4 address, if so use
+	// this.
+	//
+	// if neither, return an error, we cannot assign an router ID.
+	var routerID string
+	_, ok := cstate.Annotations[newc.LocalASN]
+	switch {
+	case ok && !net.ParseIP(cstate.Annotations[newc.LocalASN].RouterID).IsUnspecified():
+		routerID = cstate.Annotations[newc.LocalASN].RouterID
+	case !cstate.IPv4.IsUnspecified():
+		routerID = cstate.IPv4.String()
+	default:
+		return fmt.Errorf("router id not specified by annotation and no IPv4 address assigned by cilium, cannot resolve router id for virtual router with local ASN %v", newc.LocalASN)
+	}
+
+	var shouldRecreate bool
+	if localPort != bgpInfo.Global.ListenPort {
+		shouldRecreate = true
+		l.Infof("Virtual router with ASN %v local port has changed from %v to %v", newc.LocalASN, bgpInfo.Global.ListenPort, localPort)
+	}
+	if routerID != bgpInfo.Global.RouterId {
+		shouldRecreate = true
+		l.Infof("Virtual router with ASN %v router ID has changed from %v to %v", newc.LocalASN, bgpInfo.Global.RouterId, routerID)
+	}
+	if !shouldRecreate {
+		l.Debugf("No preflight reconciliation necessary for virtual router with local ASN %v", newc.LocalASN)
+		return nil
+	}
+
+	l.Infof("Recreating virtual router with ASN %v for changes to take effect", newc.LocalASN)
+
+	startReq := &gobgp.StartBgpRequest{
+		Global: &gobgp.Global{
+			Asn:        uint32(newc.LocalASN),
+			RouterId:   routerID,
+			ListenPort: localPort,
+			RouteSelectionOptions: &gobgp.RouteSelectionOptionsConfig{
+				AdvertiseInactiveRoutes: true,
+			},
+		},
+	}
+
+	// stop the old BgpServer
+	sc.Server.Stop()
```

Comment:
I haven't read gobgp code in detail, but can't we swap the new and old BGP server without peer disconnection? I think here we should have intermittent packet loss since there is a short moment when no BGP server exists. Also, when the below `NewServerWithConfig` fails for some reason, the error persists since the old server has already stopped.

This is kind of related to what I said for multiple policies issue. I prefer the "keep the existing configuration as is when new configuration has an error" strategy in here as well.

Reply:
I had hesitations starting the new BGP server while the old one exists.

If we did this, the new BGP server may have a differnt configuration. For instance, if the kubernete's router-id or local-port annotations have changed. 

We'd then have a second BGP server coming up with a slight different set of NLRIs heading to its peers. 

Maybe this is okay, but at first blush I leaned toward avoiding it. 

Happy to talk consider this further with a bit of convincing that no issues will arise from it.